File: sitevalidator/checker.py
"""Check whether URLs are reachable and return their status."""
from __future__ import annotations

import asyncio
import time
from typing import Callable, Optional

import httpx

from .models import CheckResult, Occurrence


async def check_urls(
    url_occurrences: dict[str, list[Occurrence]],
    *,
    concurrency: int = 10,
    timeout: float = 10.0,
    user_agent: str = "site-validator/0.1",
    verify_ssl: bool = True,
    on_checked: Optional[Callable[[CheckResult], None]] = None,
) -> list[CheckResult]:
    sem = asyncio.Semaphore(concurrency)
    headers = {"User-Agent": user_agent}

    async def check_one(url: str, occurrences: list[Occurrence]) -> CheckResult:
        async with sem:
            # Skip non-http paths (local file paths)
            if not url.startswith(("http://", "https://")):
                return _check_local_path(url, occurrences)

            t0 = time.monotonic()
            try:
                async with httpx.AsyncClient(
                    headers=headers,
                    timeout=httpx.Timeout(timeout),
                    follow_redirects=False,
                    verify=verify_ssl,
                ) as client:
                    # Try HEAD first (cheap), fall back to GET if server rejects it
                    resp = await client.head(url)
                    if resp.status_code == 405:
                        resp = await client.get(url)

                elapsed = (time.monotonic() - t0) * 1000
                result = CheckResult(
                    url=url,
                    status_code=resp.status_code,
                    response_time_ms=elapsed,
                    occurrences=occurrences,
                )
            except httpx.TimeoutException:
                result = CheckResult(
                    url=url,
                    error="timeout",
                    response_time_ms=(time.monotonic() - t0) * 1000,
                    occurrences=occurrences,
                )
            except httpx.ConnectError as e:
                result = CheckResult(
                    url=url,
                    error=f"connect error: {_short_error(e)}",
                    occurrences=occurrences,
                )
            except httpx.TooManyRedirects:
                result = CheckResult(
                    url=url,
                    error="too many redirects",
                    occurrences=occurrences,
                )
            except Exception as e:
                result = CheckResult(
                    url=url,
                    error=_short_error(e),
                    occurrences=occurrences,
                )

            if on_checked:
                on_checked(result)
            return result

    tasks = [
        check_one(url, occurrences)
        for url, occurrences in url_occurrences.items()
    ]
    return list(await asyncio.gather(*tasks))


def _check_local_path(url: str, occurrences: list[Occurrence]) -> CheckResult:
    """Check that a local file path exists."""
    from pathlib import Path

    path = Path(url)
    if path.exists():
        status = 200
        err = None
    else:
        # Try common index file extensions for directories
        if path.suffix == "" and (path / "index.html").exists():
            status = 200
            err = None
        else:
            status = 404
            err = None

    return CheckResult(
        url=url,
        status_code=status,
        error=err,
        occurrences=occurrences,
    )


def _short_error(e: Exception) -> str:
    msg = str(e)
    return msg[:80] if msg else type(e).__name__

File: src/checker.rs
use std::collections::HashMap;
use std::path::Path;
use std::time::Instant;

use futures::StreamExt;
use indicatif::ProgressBar;
use reqwest::{Client, StatusCode};

use crate::models::{CheckResult, Occurrence};

pub async fn check_urls(
    url_occurrences: HashMap<String, Vec<Occurrence>>,
    concurrency: usize,
    timeout_secs: f64,
    user_agent: &str,
    verify_ssl: bool,
    pb: Option<ProgressBar>,
) -> Vec<CheckResult> {
    let client = Client::builder()
        .user_agent(user_agent)
        .timeout(std::time::Duration::from_secs_f64(timeout_secs))
        .redirect(reqwest::redirect::Policy::none())
        .danger_accept_invalid_certs(!verify_ssl)
        .build()
        .expect("failed to build HTTP client");

    futures::stream::iter(url_occurrences)
        .map(|(url, occurrences)| {
            let client = client.clone();
            let pb = pb.clone();
            async move {
                let result = check_one(&client, url, occurrences).await;
                if let Some(ref pb) = pb {
                    let icon = if result.is_broken() { "✗" } else { "✓" };
                    pb.inc(1);
                    pb.set_message(format!("{} {}", icon, truncate(&result.url, 55)));
                }
                result
            }
        })
        .buffer_unordered(concurrency)
        .collect()
        .await
}

async fn check_one(client: &Client, url: String, occurrences: Vec<Occurrence>) -> CheckResult {
    if !url.starts_with("http://") && !url.starts_with("https://") {
        return check_local_path(url, occurrences);
    }

    let t0 = Instant::now();

    // Try HEAD first, fall back to GET if server rejects it
    let resp = match client.head(&url).send().await {
        Ok(r) if r.status() == StatusCode::METHOD_NOT_ALLOWED => {
            client.get(&url).send().await
        }
        other => other,
    };

    let elapsed = t0.elapsed().as_secs_f64() * 1000.0;

    match resp {
        Ok(r) => CheckResult {
            url,
            status_code: Some(r.status().as_u16()),
            error: None,
            response_time_ms: elapsed,
            occurrences,
        },
        Err(e) => {
            let error = if e.is_timeout() {
                "timeout".to_string()
            } else if e.is_connect() {
                format!("connect error: {}", short_error(&e))
            } else if e.is_redirect() {
                "too many redirects".to_string()
            } else {
                short_error(&e)
            };
            CheckResult {
                url,
                status_code: None,
                error: Some(error),
                response_time_ms: elapsed,
                occurrences,
            }
        }
    }
}

fn check_local_path(url: String, occurrences: Vec<Occurrence>) -> CheckResult {
    let path = Path::new(&url);
    let exists = path.exists()
        || path.with_extension("html").exists()
        || path.join("index.html").exists();

    CheckResult {
        url,
        status_code: Some(if exists { 200 } else { 404 }),
        error: None,
        response_time_ms: 0.0,
        occurrences,
    }
}

fn short_error(e: &reqwest::Error) -> String {
    let s = e.to_string();
    s.chars().take(80).collect()
}

fn truncate(s: &str, n: usize) -> String {
    if s.len() <= n {
        s.to_string()
    } else {
        format!("…{}", &s[s.len().saturating_sub(n - 1)..])
    }
}
